File: plugins/distortion/octaver/octaver.js
// Octaver — full-wave rectification produces a signal at 2× the input
// fundamental, perceived as one octave UP. The rectified signal is
// DC-blocked (one-pole high-pass) before mixing so it doesn't shove the
// mix off-center. Classic fuzz-octaver flavour — a single voice singing
// in 8va parallel, with the gritty edges that rectification leaves in
// the upper harmonics.

function OctaverProcessor() {
  this.prevX = 0;
  this.prevY = 0;
}
OctaverProcessor.parameterDescriptors = [
  { name: 'octave', label: 'Octave Up', defaultValue: 0.5, minValue: 0, maxValue: 1 },
  { name: 'dry',    label: 'Dry',       defaultValue: 0.7, minValue: 0, maxValue: 1 }
];
OctaverProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var octave = parameters.octave[0];
  var dry = parameters.dry[0];
  var R = 0.995;
  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    var rect = x >= 0 ? x : -x;
    var y = rect - this.prevX + R * this.prevY;
    this.prevX = rect;
    this.prevY = y;
    // ×2 compensates for the ~halved perceived amplitude after rectify.
    output[i] = x * dry + y * octave * 2;
  }
  return true;
};

registerProcessor('octaver', OctaverProcessor);

File: plugins/distortion/octaver/Octaver.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Octaver (native port) — full-wave rectification doubles the perceived
// pitch (one octave up). The rectified signal is DC-blocked by a leaky
// integrator before being mixed back under the dry voice.
public final class Octaver implements VocalMonitorNativePlugin {
    private float prevX = 0f;
    private float prevY = 0f;
    private float octave = 0.5f, dry = 0.7f;

    @Override public void init(int sr) { prevX = 0f; prevY = 0f; }
    @Override public String[] parameterNames() { return new String[] { "octave", "dry" }; }
    @Override public float parameterMin(String n) { return 0f; }
    @Override public float parameterMax(String n) { return 1f; }
    @Override public float parameterDefault(String n) { return "octave".equals(n) ? 0.5f : 0.7f; }
    @Override public String parameterLabel(String n) { return "octave".equals(n) ? "Octave Up" : "Dry"; }
    @Override public void setParameter(String n, float v) {
        if ("octave".equals(n)) octave = v;
        else if ("dry".equals(n)) dry = v;
    }

    @Override
    public void process(float[] input, float[] output) {
        final float R = 0.995f;
        final float octGain = octave * 2f;
        final float dryLocal = dry;
        float pX = prevX, pY = prevY;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            float rect = x >= 0 ? x : -x;
            float y = rect - pX + R * pY;
            pX = rect; pY = y;
            output[i] = x * dryLocal + y * octGain;
        }
        prevX = pX; prevY = pY;
    }
}
